### ISSUE TITLE
error: output of --print=file-names missing when learning about target-specific information from rustc when `--target` is specified

### ISSUE THREAD

**Issue by zjp-CN** [obi1kenobi/cargo-semver-checks#1068]
### Steps to reproduce the bug with the above code

Thanks for this great tool!


I see a bunch of these repos whos semver-checks results on **non-** x86_64-unknown-linux-gnu targets report as the title shows.
One of the repos is https://github.com/arceos-org/arm_gicv2 . (I can list more if needed.)

Reproducible commands:

```sh
# latest hash: https://github.com/arceos-org/arm_gicv2/commit/cf756f7623b6b6af2f6fe423d1da19ebe96e7e1d
git clone https://github.com/arceos-org/arm_gicv2.git 
cd arm_gicv2
rustup target add aarch64-unknown-none-softfloat
cargo semver-checks --target aarch64-unknown-none-softfloat
```

### Actual Behaviour

I got compilation error from rustc.

```rust
error: output of --print=file-names missing when learning about target-specific information from rustc
command was: `/root/.cargo/bin/sccache /root/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/rustc - --crate-name ___ --print=file-names --cap-lints=allow --target aarch64-unknown-none-softfloat --crate-type bin --crate-type rlib --crate-type dylib --crate-type cdylib --crate-type staticlib --crate-type proc-macro --print=sysroot --print=split-debuginfo --print=crate-name --print=cfg`

--- stdout
___
lib___.rlib
lib___.a
/root/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu
off
___
debug_assertions
fmt_debug="full"
overflow_checks
panic="abort"
proc_macro
relocation_model="static"
target_abi="softfloat"
target_arch="aarch64"
target_endian="little"
target_env=""
target_has_atomic
target_has_atomic="128"
target_has_atomic="16"
target_has_atomic="32"
target_has_atomic="64"
target_has_atomic="8"
target_has_atomic="ptr"
target_has_atomic_equal_alignment="128"
target_has_atomic_equal_alignment="16"
target_has_atomic_equal_alignment="32"
target_has_atomic_equal_alignment="64"
target_has_atomic_equal_alignment="8"
target_has_atomic_equal_alignment="ptr"
target_has_atomic_load_store
target_has_atomic_load_store="128"
target_has_atomic_load_store="16"
target_has_atomic_load_store="32"
target_has_atomic_load_store="64"
target_has_atomic_load_store="8"
target_has_atomic_load_store="ptr"
target_os="none"
target_pointer_width="64"
target_vendor="unknown"
ub_checks


-----

error: failed to build rustdoc for crate arm_gicv2 v0.1.0
note: this is usually due to a compilation error in the crate,
      and is unlikely to be a bug in cargo-semver-checks
note: the following command can be used to reproduce the compilation error:
      cargo new --lib example &&
          cd example &&
          echo '[workspace]' >> Cargo.toml &&
          cargo add --path /rust/tmp/os-checker/repos/arm_gicv2 &&
          cargo check

error: aborting due to failure to build rustdoc for crate arm_gicv2 v0.1.0
```

I also run the suggested commands to reproduce the compilation error, and it compiles just fine.

### Expected Behaviour

There should be no compilation errors.

### Generated System Information

```
#### Software version

cargo-semver-checks 0.38.0 (507393d-modified)

#### Operating system

- OS: Linux (Ubuntu 24.04)
- Kernel: 6.8.0-51-generic

#### Command-line                                                                                                                                                                                                                                                                                                                                           ```bash                                                                                                                                                                       /root/.cargo/bin/cargo-semver-checks semver-checks --bugreport                                                                                                                ```                                                                                                                                                                                                                                                                                                                                                         #### cargo version                                                                                                                                                                                                                                                                                                                                          ```
> cargo -V
cargo 1.85.0-nightly (c86f4b3a1 2024-12-24)
```

#### Compile time information

- Profile: release
- Target triple: x86_64-unknown-linux-gnu
- Family: unix
- OS: linux
- Architecture: x86_64
- Pointer width: 64
- Endian: little
- CPU features: fxsr,sse,sse2,x87
- Host: x86_64-unknown-linux-gnu

Cargo build configuration:
--------------------------
[build]
rustc-wrapper = "/root/.cargo/bin/sccache"
```

### Build Configuration

None.

### Additional Context

I think it might be a bug with RUSTCFLAGS. When I look into the generated cargo doc command from semver-checks, it's like the following

```rust
export RUSTC_BOOTSTRAP=1
export RUSTDOCFLAGS="-Z unstable-options --document-private-items --document-hidden-items --output-format=json --cap-lints=allow"
export RUSTFLAGS=--cap-lints=allow
cargo doc --manifest-path target/semver-checks/local-arm_gicv2-0_1_0-01666ec060466c14/Cargo.toml --target-dir arm_gicv2/target/semver-checks/local-arm_gicv2-0_1_0-01666ec060466c14/target --package arm_gicv2@0.1.0 --lib --target aarch64-unknown-none-softfloat --no-deps --color=always
```

This just results in the exact error report above. (I'm not sure if the hash string in path is the same for your local testing.)

And when `RUSTFLAGS=--cap-lints=allow` is not included, it compiles and successfully generates the doc.

So I try removing [RUSTFLAGS from `run_cargo_doc` fn](https://github.com/obi1kenobi/cargo-semver-checks/blob/507393d1fb1ff1841dcd36a7b2acffeb8e811b44/src/data_generation/generate.rs#L250-L256), install it locally, and run `cargo semver-checks --target aarch64-unknown-none-softfloat` agian, semver-checks just works.


```rust
    Building arm_gicv2 v0.1.0 (current)
       Built [   0.733s] (current)
     Parsing arm_gicv2 v0.1.0 (current)
      Parsed [   0.002s] (current)
    Building arm_gicv2 v0.1.0 (baseline)
        Built [   0.558s] (baseline)
      Parsing arm_gicv2 v0.1.0 (baseline)
       Parsed [   0.002s] (baseline)
     Checking arm_gicv2 v0.1.0 -> v0.1.0 (no change)
      Checked [   0.020s] 120 checks: 120 pass, 0 skip
      Summary no semver update required
     Finished [   3.697s] arm_gicv2
```

**Comment by obi1kenobi** [comment]
Sorry for the delay, I was travelling and only just got home.

As far as I can tell, this is a bug upstream, likely in `rustc` or `rustdoc` 😮 `--cap-lints=allow` means "please don't error out on lints". It has no effect on the compiled code or anything related to the target, so specifying that flag should never cause code to not compile. What you've found is quite shocking to me!

Would you mind making a bug report issue in https://github.com/rust-lang/rust and linking it to this one?

If you'd prefer not to, I can also open the issue upstream. But since you already did all the work to narrow down the source of the bug, I didn't want to appear to take credit for your work by opening the upstream issue :)

**Comment by zjp-CN** [comment]
😄  I'd like to open one on rust-lang repo. I'll try to make reproducible code that can be run without cargo-semver-checks.

**Comment by zjp-CN** [comment]
I've submitted it: https://github.com/rust-lang/rust/issues/135233

**Comment by weihanglo** [comment]
I believe https://github.com/rust-lang/cargo/pull/15036 should fix this. It will get into the main rust-lang/rust repo on Friday, so expect to see it on nightly in the weekend!

**Comment by zjp-CN** [comment]
Well, I think https://github.com/rust-lang/cargo/pull/15036 doesn't solve the issue here.

I've tested with nightly-2025-01-12 where that cargo PR landed, the same `error: output of --print=file-names missing when learning about target-specific information from rustc` occurs. You can see them from the [CI result](https://github.com/os-checker/dockerfiles/actions/runs/12730049370/job/35482566652) which runs on nightly-2025-01-01 and nightly-2025-01-12[^1].

[^1]: Interestingly, nightly-2025-01-12 actually makes `RUSTFLAGS="-Dwarnings -Awarnings" cargo check` pass while that check cmd fails on nightly-2025-01-01. So a fix for semver-checks is to append -Awarnings in RUSTFLAGS, but it's too subtle to be worth considering.

I believe --cap-lint=allow fix introduced in handling -Dwarnings (ref https://github.com/obi1kenobi/cargo-semver-checks/issues/589 )  is not proper. --cap-lint=allow eliminates warnings even if cargo sets -Awarnings in the cargo PR. cargo-semver-checks should remove -Dwarnings from RUSTFLAGS and do not set --cap-lint.

**Comment by weihanglo** [comment]
Oh I see now. For `--cap-lints`, the behavior for this flag is the first occurrence wins <https://github.com/rust-lang/rust/issues/54282>. I am unsure if we really want to and can depend on it.

**Comment by obi1kenobi** [comment]
My preference, if at all possible, would be to stick to `--cap-lints=allow` since we're using the flag for exactly its intended purpose: declaring that we don't care about lints, at all, under any circumstance. By comparison, editing the *user's* provided `RUSTFLAGS` feels icky and wrong.

That said, I understand that the cargo -> rustc interface is limited and there are workarounds in place. I trust the judgment of cargo team members like @weihanglo on whether it's feasible to make `--cap-lints=allow` work for us in the near term, or if we should go with editing user-provided `RUSTFLAGS` instead.

**Comment by obi1kenobi** [comment]
Has one of you had a chance to test whether the attempted fix for this in recent nightly Rust is working as expected? I see the [cargo issue](https://github.com/rust-lang/rust/issues/116626) is closed as completed, and its linked PR is merged and likely available in nightly. Would be great to validate it!

I'm waiting on a rustdoc JSON change before I release a new `cargo-semver-checks` version, and hoping to make it in the next 3-5 days or so. I'd love to know if the recommended workaround is "please use very recent Rust" or if it should be something else.

**Comment by weihanglo** [comment]
For passing `--cap-lints` it hasn't. I got flu this week so is in a slow mode. Will ping people on Zulip asking https://github.com/rust-lang/rust/issues/54282

**Comment by obi1kenobi** [comment]
Oh no, I'm sorry to hear it. Hope you feel better soon! Please take care and prioritize your health, that's much more important than this issue.

**Comment by zjp-CN** [comment]
@weihanglo Awww, that’s terrible. Take care and get lots of rest. ❤ 

Don't worry. I've [started a thread](https://rust-lang.zulipchat.com/#narrow/channel/246057-t-cargo/topic/RUSTFLAGS.3D--cap-lints.3Dallow.20suppresses.20crate-type.20info) on zulichat's Cargo channel.


This issue is less urgent for me, because I made a fork of cargo-semver-checks to solve the issue for now.

Thank you all for trying to handle this!

**Comment by tgross35** [comment]
For the time being, is there any kind of workaround to run cargo-semver-checks on non-native targets?

**Comment by tgross35** [comment]
Just read some more at https://github.com/obi1kenobi/cargo-semver-checks-action/issues/98 it looks like `--print=supported-crate-types` is unstably available. Could cargo-semver-checks get an unstable flag that checks `RUSTC_BOOTSTRAP=1 rustc +stable --print=supported-crate-types -Zunstable-options` rather than relying on the output here?

**Comment by obi1kenobi** [comment]
`cargo-semver-checks` isn't making that call at all, unfortunately. This is something that IIUC happens under the hood as a result of passing `--cap-lints=allow` together with `--target` to `cargo doc`.

There's no good way for us to avoid `--cap-lints=allow`, because otherwise e.g. a user's `-D warnings` setting (in env vars or other config) will prevent us from building rustdoc JSON. The role of `--cap-lints=allow` is precisely for use cases like ours — where the lint is irrelevant — so we're using it as designed.

The issue is that `--cap-lints=allow` is broken right now. So this is entirely blocked upstream from us on this: https://github.com/rust-lang/rust/issues/138640#issuecomment-2755743872

I'm hopeful that can be resolved upstream in a timely fashion. I'm really wary of anything that might increase the support burden of `cargo-semver-checks`, which is the main argument against adding a `cargo-semver-checks` option to disable passing `--cap-lints` — it risks introducing more ways for things to interact poorly, which will require more issue triage and debugging.

**Comment by obi1kenobi** [comment]
> For the time being, is there any kind of workaround to run cargo-semver-checks on non-native targets?

You might be able to _partially_ recover `cargo-semver-checks` functionality, via a workflow that's generally discouraged otherwise because we _know_ it will be broken sooner or later. Since `cargo doc` is the command that fails, you could generate the baseline + current rustdoc JSON manually and then use CLI flags to ask `cargo-semver-checks` to compare the two rustdoc JSON files.

This won't load `cargo metadata` information, obviously, so you won't get the benefit of any [manifest lints like breaking changes to package versions](https://predr.ag/blog/breakage-in-the-cargo-toml-how-rust-package-features-work/), for example. But most of the lints should work fine.

[Here's the command](https://github.com/obi1kenobi/cargo-semver-checks/blob/030af2032e93a64a6a40c4deaa0f57f262042426/src/data_generation/generate.rs#L241-L297) we use to generate rustdoc JSON. It's a bit involved, sorry about that :)

**Comment by tgross35** [comment]
Thanks for the info! I'm happy to accept a hack here if it gets things off the ground for checking other targets, better than no checks at all :)
